azureml/services.py

```diff
@@ -615,67 +616,78 @@ def _get_dataframe_schema(function):
         if args.keywords is not None:
-            all_args.append('**' + args.keywords)
+            all_args.append(u'**' + args.keywords)
     
         # pass position arguments...
-        main_source += ', '.join(all_args)
-        main_source += '))' + chr(10)
+        main_source += u', '.join(all_args)
+        main_source += u'))' + chr(10)
     
     ret_annotation = _get_annotation('return', function)
     if _get_dataframe_schema(function):
         # function just returns a data frame directly
-        main_source += '    if len(results) == 1:' + chr(10)
-        main_source += '        return results[0]' + chr(10)
-        main_source += '    return pandas.DataFrame(results)' + chr(10)
+        main_source += u'    if len(results) == 1:' + chr(10)
+        main_source += u'        return results[0]' + chr(10)
+        main_source += u'    return pandas.DataFrame(results)' + chr(10)
     elif isinstance(ret_annotation, tuple):
         # multi-value return support...
         format = []
         arg_names = []
         for index, ret_type in enumerate(ret_annotation):
-            arg_names.append('r' + str(index))
+            arg_names.append(u'r' + str(index))
             t = _annotation_to_type(ret_type)
             if t == OBJECT_NAME:
-                format.append('_encode(r' + str(index) + ')')
+                format.append(u'_encode(r' + str(index) + u')')
             else:
-                format.append('r' + str(index))
-        main_source += '    return pandas.DataFrame([(' + ', '.join(format) + ') for ' + ', '.join(arg_names) + ' in results])' + chr(10)
+                format.append(u'r' + str(index))
+        main_source += u'    return pandas.DataFrame([(' + u', '.join(format) + u') for ' + ', '.join(arg_names) + u' in results])' + chr(10)
     elif _get_arg_type('return', function) == OBJECT_NAME:
-        main_source += '    return pandas.DataFrame([_encode(r) for r in results])' + chr(10)
+        main_source += u'    return pandas.DataFrame([_encode(r) for r in results])' + chr(10)
     else:
-        main_source += '    return pandas.DataFrame(results)' + chr(10)
+        main_source += u'    return pandas.DataFrame(results)' + chr(10)
     
     return main_source
 
 def _get_source(function):
     source_file = inspect.getsourcefile(function)
+    encoding = ''
     try:
-        source_text = ''.join(open(source_file).readlines())
+        with open(source_file, 'rb') as source_file:
+            line1 = source_file.readline()
+            line2 = source_file.readline()
+            if line1[:3] == '\xef\xbb\xbf':
+                encoding = 'utf-8-sig'
+            else:
+                match = re.search(b"coding[:=]\s*([-\w.]+)", line1) or re.search(b"coding[:=]\s*([-\w.]+)", line2)
+                if match:
+                    encoding = match.groups()[0]
+        with codecs.open(source_file, 'r', encoding) as source_file:
+            source_text = u''.join(source_file.readlines())
```

Comment:
Any reason you don't just use source_file.read() here? The only way this should actually be any more efficient would be to keep the entire source in a list and then join it at the end.